File: kirun-js/__tests__/engine/function/system/date/AddSubtractTimeTest.ts
import { DateTime, Settings } from 'luxon';
import { KIRunFunctionRepository, KIRunSchemaRepository } from '../../../../../src';
import { AbstractDateFunction } from '../../../../../src/engine/function/system/date/AbstractDateFunction';
import { AddSubtractTime } from '../../../../../src/engine/function/system/date/AddSubtractTime';
import { FunctionExecutionParameters } from '../../../../../src/engine/runtime/FunctionExecutionParameters';

const addFunction = new AddSubtractTime(true);
const subtractFunction = new AddSubtractTime(false);

Settings.defaultZone = 'Asia/Kolkata';

describe('AddSubtractTime', () => {
    test('should add time with no values', async () => {
        const fep = new FunctionExecutionParameters(
            new KIRunFunctionRepository(),
            new KIRunSchemaRepository(),
        ).setArguments(
            new Map<string, any>([
                [AbstractDateFunction.PARAMETER_TIMESTAMP_NAME, '2025-01-01T10:20:35+05:30'],
            ]),
        );

        const result = await addFunction.execute(fep);

        expect(
            result.allResults()[0].getResult().get(AbstractDateFunction.EVENT_TIMESTAMP_NAME),
        ).toBe('2025-01-01T10:20:35.000+05:30');
    });

    test('shoudl subtract time with no values', async () => {
        const fep = new FunctionExecutionParameters(
            new KIRunFunctionRepository(),
            new KIRunSchemaRepository(),
        ).setArguments(
            new Map<string, any>([
                [AbstractDateFunction.PARAMETER_TIMESTAMP_NAME, '2025-01-01T10:20:35+05:30'],
            ]),
        );

        const result = await subtractFunction.execute(fep);

        expect(
            result.allResults()[0].getResult().get(AbstractDateFunction.EVENT_TIMESTAMP_NAME),
        ).toBe('2025-01-01T10:20:35.000+05:30');
    });

    test('should add time with all values', async () => {
        const fep = new FunctionExecutionParameters(
            new KIRunFunctionRepository(),
            new KIRunSchemaRepository(),
        ).setArguments(
            new Map<string, any>([
                [AbstractDateFunction.PARAMETER_TIMESTAMP_NAME, '2025-01-01T10:20:35+05:30'],
                [AddSubtractTime.PARAMETER_YEARS_NAME, 1],
                [AddSubtractTime.PARAMETER_MONTHS_NAME, 1],
                [AddSubtractTime.PARAMETER_DAYS_NAME, 1],
                [AddSubtractTime.PARAMETER_HOURS_NAME, 1],
                [AddSubtractTime.PARAMETER_MINUTES_NAME, 1],
                [AddSubtractTime.PARAMETER_SECONDS_NAME, 1],
            ]),
        );

        const result = await addFunction.execute(fep);

        expect(
            result.allResults()[0].getResult().get(AbstractDateFunction.EVENT_TIMESTAMP_NAME),
        ).toBe('2026-02-02T11:21:36.000+05:30');
    });

    test('should subtract time with all values', async () => {
        const fep = new FunctionExecutionParameters(
            new KIRunFunctionRepository(),
            new KIRunSchemaRepository(),
        ).setArguments(
            new Map<string, any>([
                [AbstractDateFunction.PARAMETER_TIMESTAMP_NAME, '2025-01-01T10:20:35+05:30'],
                [AddSubtractTime.PARAMETER_YEARS_NAME, 1],
                [AddSubtractTime.PARAMETER_MONTHS_NAME, 1],
                [AddSubtractTime.PARAMETER_DAYS_NAME, 1],
                [AddSubtractTime.PARAMETER_HOURS_NAME, 1],
                [AddSubtractTime.PARAMETER_MINUTES_NAME, 1],
                [AddSubtractTime.PARAMETER_SECONDS_NAME, 1],
                [AddSubtractTime.PARAMETER_MILLISECONDS_NAME, 8],
            ]),
        );

        const result = await subtractFunction.execute(fep);

        expect(
            result.allResults()[0].getResult().get(AbstractDateFunction.EVENT_TIMESTAMP_NAME),
        ).toBe('2023-11-30T09:19:33.992+05:30');
    });
});

File: kirun-java/src/test/java/com/fincity/nocode/kirun/engine/function/system/date/AddSubtractTimeTest.java
package com.fincity.nocode.kirun.engine.function.system.date;

import java.util.Map;
import java.util.TimeZone;

import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;

import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveFunctionRepository;
import com.fincity.nocode.kirun.engine.repository.reactive.KIRunReactiveSchemaRepository;
import com.fincity.nocode.kirun.engine.runtime.reactive.ReactiveFunctionExecutionParameters;
import com.google.gson.JsonPrimitive;

import reactor.test.StepVerifier;

class AddSubtractTimeTest {

    @BeforeAll
    public static void setup() {
        TimeZone.setDefault(TimeZone.getTimeZone("UTC"));
    }

    @Test
    void testAddTimeWithNoArguments() {
        AddSubtractTime addSubtractTime = new AddSubtractTime(true);
        ReactiveFunctionExecutionParameters parameters = new ReactiveFunctionExecutionParameters(
                new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
                .setArguments(
                        Map.of(AddSubtractTime.PARAMETER_TIMESTAMP_NAME,
                                new JsonPrimitive("2024-11-11T14:55:00+05:30")));

        StepVerifier.create(addSubtractTime.execute(parameters)
                .map(functionOutput -> functionOutput.allResults().get(0).getResult()
                        .get(AddSubtractTime.EVENT_TIMESTAMP_NAME).getAsString()))
                .expectNext("2024-11-11T14:55:00+05:30")
                .verifyComplete();
    }

    @Test
    void testAddTimeSomeUnits() {

        AddSubtractTime addSubtractTime = new AddSubtractTime(true);
        ReactiveFunctionExecutionParameters parameters = new ReactiveFunctionExecutionParameters(
                new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
                .setArguments(
                        Map.of(AddSubtractTime.PARAMETER_TIMESTAMP_NAME,
                                new JsonPrimitive("2024-11-11T14:55:00.000+05:30"),
                                AddSubtractTime.PARAMETER_YEARS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_MONTHS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_DAYS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_HOURS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_MINUTES_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_SECONDS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_MILLISECONDS_NAME, new JsonPrimitive(33)));

        StepVerifier.create(addSubtractTime.execute(parameters)
                .map(functionOutput -> functionOutput.allResults().get(0).getResult()
                        .get(AddSubtractTime.EVENT_TIMESTAMP_NAME).getAsString()))
                .expectNext("2025-12-12T15:56:01.033+05:30")
                .verifyComplete();
    }

    @Test
    void testSubtractTimeSomeUnits() {

        AddSubtractTime addSubtractTime = new AddSubtractTime(false);
        ReactiveFunctionExecutionParameters parameters = new ReactiveFunctionExecutionParameters(
                new KIRunReactiveFunctionRepository(), new KIRunReactiveSchemaRepository())
                .setArguments(
                        Map.of(AddSubtractTime.PARAMETER_TIMESTAMP_NAME,
                                new JsonPrimitive("2024-11-11T14:55:00+05:30"),
                                AddSubtractTime.PARAMETER_MINUTES_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_SECONDS_NAME, new JsonPrimitive(1),
                                AddSubtractTime.PARAMETER_MILLISECONDS_NAME, new JsonPrimitive(33)));

        StepVerifier.create(addSubtractTime.execute(parameters)
                .map(functionOutput -> functionOutput.allResults().get(0).getResult()
                        .get(AddSubtractTime.EVENT_TIMESTAMP_NAME).getAsString()))
                .expectNext("2024-11-11T14:53:58.967+05:30")
                .verifyComplete();
    }
}
